### PR TITLE
Port 0.6 runtime to DukTape [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by davidkel**
Signed-off-by: Dave Kelsey <d_kelsey@uk.ibm.com>

### CODE REVIEW

**Comment by davidkel** [comment]
Damn, the build still failed with the same problem and it should be using the latest duktape.
```
  1) Transaction (participant specific) system tests should submit and execute a transaction that updates a new participant with a relationship in an participant registry:
     Error: Error:Failed to execute transaction or query(Timeout expired while executing transaction)
      at TransactionContext.transactionContext.on (/home/travis/build/hyperledger/composer/packages/composer-connector-hlf/lib/hfcutil.js:91:28)
      at Object.callback (/home/travis/build/hyperledger/composer/node_modules/hfc/lib/hfc.js:1807:42)
      at /home/travis/build/hyperledger/composer/node_modules/hfc/node_modules/grpc/src/node/src/client.js:422:14
```

**Comment by davidkel** [comment]
run it locally and got a crash in the middle of the tests
```
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x7facd0fe81f8 pc=0x7fadd5dfc438]

runtime stack:
runtime.throw(0x976af5, 0x2a)
	/opt/go/src/runtime/panic.go:566 +0x95
runtime.sigpanic()
	/opt/go/src/runtime/sigpanic_unix.go:12 +0x2cc

goroutine 3948 [syscall, locked to thread]:
runtime.cgocall(0x834b63, 0xc420e6faf0, 0x0)
	/opt/go/src/runtime/cgocall.go:131 +0x110 fp=0xc420e6fac0 sp=0xc420e6fa80
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3._cgo_cmalloc(0x86a, 0x0)
	build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/_obj/_cgo_gotypes.go:3459 +0x4a fp=0xc420e6faf0 sp=0xc420e6fac0
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3._Cfunc_CString(0xc4201e8400, 0x869, 0x0)
	build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/_obj/_cgo_gotypes.go:130 +0x33 fp=0xc420e6fb20 sp=0xc420e6faf0
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.(*Context).PushString(0xc420d0df00, 0xc4201e8400, 0x869, 0xc420bab700, 0x1)
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/api.go:1072 +0x39 fp=0xc420e6fb80 sp=0xc420e6fb20
main.(*DataCollection).get(0xc420c0a210, 0xc420d0df00, 0x0)
	/opt/gopath/src/build-chaincode/datacollection.go:142 +0x2f0 fp=0xc420e6fc98 sp=0xc420e6fb80
main.(*DataCollection).(main.get)-fm(0xc420d0df00, 0x7fadd0917c00)
	/opt/gopath/src/build-chaincode/datacollection.go:62 +0x34 fp=0xc420e6fcc0 sp=0xc420e6fc98
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.goFunctionCall(0x7fadd0005190, 0xc420e6fd60)
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/duktape.go:118 +0xfc fp=0xc420e6fd00 sp=0xc420e6fcc0
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3._cgoexpwrap_b10f6b0ce12e_goFunctionCall(0x7fadd0005190, 0x3)
	build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/_obj/_cgo_gotypes.go:3436 +0x2b fp=0xc420e6fd20 sp=0xc420e6fd00
runtime.call32(0x0, 0x7fadd557b2a8, 0x7fadd557b348, 0x10)
	/opt/go/src/runtime/asm_amd64.s:479 +0x4c fp=0xc420e6fd50 sp=0xc420e6fd20
runtime.cgocallbackg1(0x0)
	/opt/go/src/runtime/cgocall.go:283 +0x19d fp=0xc420e6fdc8 sp=0xc420e6fd50
runtime.cgocallbackg(0x0)
	/opt/go/src/runtime/cgocall.go:170 +0x84 fp=0xc420e6fe30 sp=0xc420e6fdc8
runtime.cgocallback_gofunc(0x421b68, 0x835f61, 0xc420e6fee0, 0xc420024800)
	/opt/go/src/runtime/asm_amd64.s:728 +0x74 fp=0xc420e6fe50 sp=0xc420e6fe30
runtime.asmcgocall(0x835f61, 0xc420e6fee0)
	/opt/go/src/runtime/asm_amd64.s:575 +0x42 fp=0xc420e6fe58 sp=0xc420e6fe50
runtime.cgocall(0x835f61, 0xc420e6fee0, 0x7fad00000000)
	/opt/go/src/runtime/cgocall.go:132 +0x128 fp=0xc420e6fe98 sp=0xc420e6fe58
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3._Cfunc_duk_pcall(0x7fadd0005190, 0x7fad00000000, 0x0)
	build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/_obj/_cgo_gotypes.go:2207 +0x4d fp=0xc420e6fee0 sp=0xc420e6fe98
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.(*Context).Pcall(0xc420d0dec0, 0x0, 0x6)
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/api.go:773 +0x85 fp=0xc420e6ff28 sp=0xc420e6fee0
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.setTimeout.func1(0x3ff0000000000000, 0xc420d0dec0, 0x40a0780000000000)
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/timers.go:55 +0x12a fp=0xc420e6ff88 sp=0xc420e6ff28
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc420e6ff90 sp=0xc420e6ff88
created by build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.setTimeout
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/timers.go:58 +0xab

goroutine 1 [chan receive]:
build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.chatWithPeer(0xc420010077, 0x40, 0x7fadd67f5d40, 0xc420015c90, 0xff66a0, 0xc420022030, 0x0, 0x0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/chaincode.go:229 +0x410
build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.Start(0xff66a0, 0xc420022030, 0x1, 0x1)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/chaincode.go:103 +0x943
main.main()
	/opt/gopath/src/build-chaincode/main.go:36 +0x104

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 12 [IO wait]:
net.runtime_pollWait(0x7fadd67f5b20, 0x72, 0x3)
	/opt/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(0xc42000ea70, 0x72, 0xc420030b18, 0xc420014160)
	/opt/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitRead(0xc42000ea70, 0xff27e0, 0xc420014160)
	/opt/go/src/net/fd_poll_runtime.go:78 +0x34
net.(*netFD).Read(0xc42000ea10, 0xc42017c000, 0x8000, 0x8000, 0x0, 0xff27e0, 0xc420014160)
	/opt/go/src/net/fd_unix.go:243 +0x1a1
net.(*conn).Read(0xc4200221e0, 0xc42017c000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/opt/go/src/net/net.go:173 +0x70
bufio.(*Reader).fill(0xc4200109c0)
	/opt/go/src/bufio/bufio.go:97 +0x10c
bufio.(*Reader).Read(0xc4200109c0, 0xc4200ca278, 0x9, 0x9, 0x0, 0x46e310, 0x448b71)
	/opt/go/src/bufio/bufio.go:209 +0x1bc
io.ReadAtLeast(0xfeed60, 0xc4200109c0, 0xc4200ca278, 0x9, 0x9, 0x9, 0x68ed8a, 0xc42015a788, 0xc420010ba0)
	/opt/go/src/io/io.go:307 +0xa4
io.ReadFull(0xfeed60, 0xc4200109c0, 0xc4200ca278, 0x9, 0x9, 0x30, 0xc42015a788, 0xfef960)
	/opt/go/src/io/io.go:325 +0x58
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/golang.org/x/net/http2.readFrameHeader(0xc4200ca278, 0x9, 0x9, 0xfeed60, 0xc4200109c0, 0x0, 0xc400000000, 0x8b2, 0x8b2)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/golang.org/x/net/http2/frame.go:237 +0x7b
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/golang.org/x/net/http2.(*Framer).ReadFrame(0xc4200ca240, 0xff2e20, 0xc420c0a2d0, 0x0, 0x0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/golang.org/x/net/http2/frame.go:464 +0xa4
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.(*framer).readFrame(0xc42015a630, 0xc420c0a2d0, 0xc420c0a2d0, 0x0, 0x0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/http_util.go:430 +0x2f
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.(*http2Client).reader(0xc4201022d0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/http2_client.go:820 +0xb6
created by build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.newHTTP2Client
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/http2_client.go:167 +0x89f

goroutine 13 [select]:
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.(*http2Client).controller(0xc4201022d0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/http2_client.go:898 +0x597
created by build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.newHTTP2Client
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/http2_client.go:197 +0xdd0

goroutine 14 [select]:
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*addrConn).transportMonitor(0xc42006c000)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/clientconn.go:623 +0x36a
created by build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*ClientConn).newAddrConn
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/clientconn.go:409 +0x447

goroutine 16 [select]:
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.NewClientStream.func1(0xff8dc0, 0xc4201022d0, 0xc4201023c0, 0xc42006c480)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/stream.go:186 +0x393
created by build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.NewClientStream
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/stream.go:203 +0x96d

goroutine 18 [select]:
build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.chatWithPeer.func1(0xc420010d80, 0xc420108130, 0x7fadd67f5d40, 0xc420015c90)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/chaincode.go:185 +0xe05
created by build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.chatWithPeer
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/chaincode.go:228 +0x3ed

goroutine 3951 [select]:
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc42015a7b0, 0xc4200fa930, 0x5, 0x5, 0x0, 0x0, 0x0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/transport.go:142 +0x5ed
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport.(*Stream).Read(0xc4201023c0, 0xc4200fa930, 0x5, 0x5, 0x41bab5, 0xc4200fb340, 0xc420d0dec8)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/transport/transport.go:301 +0x56
io.ReadAtLeast(0xfef860, 0xc4201023c0, 0xc4200fa930, 0x5, 0x5, 0x5, 0x7fadd037cf50, 0xc420109e20, 0x7fadd037cf50)
	/opt/go/src/io/io.go:307 +0xa4
io.ReadFull(0xfef860, 0xc4201023c0, 0xc4200fa930, 0x5, 0x5, 0x475f2c, 0x7fadd0005190, 0x7)
	/opt/go/src/io/io.go:325 +0x58
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*parser).recvMsg(0xc4200fa920, 0x300000002, 0x1, 0xc4202125d8, 0x1000000004517c1, 0xc42007c9c0, 0x200000003)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/rpc_util.go:231 +0x61
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.recv(0xc4200fa920, 0xff6b20, 0x102ba50, 0xc4201023c0, 0x0, 0x0, 0x919760, 0xc420e0d860, 0xc42007c9c0, 0xc420055c00)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/rpc_util.go:312 +0x40
build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc.(*clientStream).RecvMsg(0xc42006c480, 0x919760, 0xc420e0d860, 0x0, 0x0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/vendor/google.golang.org/grpc/stream.go:291 +0xbb
build-chaincode/vendor/github.com/hyperledger/fabric/protos.(*chaincodeSupportRegisterClient).Recv(0xc420015c90, 0x66, 0xc420ada2a0, 0xc42021f1e0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/protos/chaincode.pb.go:490 +0x62
build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.chatWithPeer.func1.1(0xc420108130, 0x7fadd67f5d40, 0xc420015c90, 0xc420010de0)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/chaincode.go:181 +0x35
created by build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.chatWithPeer.func1
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/chaincode.go:183 +0xe43

goroutine 3946 [chan receive]:
main.(*Concerto).Invoke(0xc4200fb1c0, 0xffb520, 0xc420220e40, 0xc420b56360, 0x14, 0xc420baa370, 0x0, 0x0, 0x0, 0x0, ...)
	/opt/gopath/src/build-chaincode/concerto.go:150 +0x377
main.(*Chaincode).Invoke(0xc420022030, 0xffb520, 0xc420220e40, 0xc420b56360, 0x14, 0xc420baa370, 0x0, 0x0, 0x0, 0x0, ...)
	/opt/gopath/src/build-chaincode/chaincode.go:54 +0x34e
build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.(*Handler).handleTransaction.func1(0xc42000ef50, 0xc420e0c820)
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/handler.go:295 +0x4ac
created by build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim.(*Handler).handleTransaction
	/opt/gopath/src/build-chaincode/vendor/github.com/hyperledger/fabric/core/chaincode/shim/handler.go:311 +0x49

goroutine 3950 [semacquire]:
sync.runtime_Semacquire(0xc4200fb224)
	/opt/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc4200fb220)
	/opt/go/src/sync/mutex.go:85 +0xd0
build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.setTimeout.func1(0x3ff0000000000000, 0xc420d0def0, 0x40a07a0000000000)
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/timers.go:45 +0x72
created by build-chaincode/vendor/gopkg.in/olebedev/go-duktape%2ev3.setTimeout
	/opt/gopath/src/build-chaincode/vendor/gopkg.in/olebedev/go-duktape.v3/timers.go:58 +0xab
```
different stack trace to the one Simon saw

**Comment by davidkel** [comment]
run locally again and got this error
```
  1) Participant system tests should resolve participants containing participant relationships from a participant registry:
     Error: Error:Failed to execute transaction or query(Timeout expired while executing transaction)
      at TransactionContext.transactionContext.on (/home/vagrant/src/fabric-composer/packages/composer-connector-hlf/lib/hfcutil.js:91:28)
      at Object.callback (/home/vagrant/src/fabric-composer/node_modules/hfc/lib/hfc.js:1807:42)
      at /home/vagrant/src/fabric-composer/node_modules/hfc/node_modules/grpc/src/node/src/client.js:422:14
```

File: .gitmodules [hunk 1]
```diff
@@ -5,9 +5,9 @@
 [submodule "packages/composer-runtime-hlf/vendor/gopkg.in/sourcemap.v1"]
```
Comment (sstone1) [inline .gitmodules]: We can also remove sourcemap.v1 as it is there as it is a dependency of otto.

**Comment by davidkel** [comment]
Experimenting to test out if the code change stops the crash, however the code change is not how the problem should be solved but a driver to determine what the problem might be. Fixes the crash locally now lets see if it fixes the build.

**Comment by davidkel** [comment]
A different failure (still can't recreate locally)
```
93 passing (7m)
  1 failing
  1) Asset system tests should update an asset in an asset registry:
     Error: Error:Failed to execute transaction or query(Timeout expired while executing transaction)
      at TransactionContext.transactionContext.on (/home/travis/build/hyperledger/composer/packages/composer-connector-hlf/lib/hfcutil.js:91:28)
      at Object.callback (/home/travis/build/hyperledger/composer/node_modules/hfc/lib/hfc.js:1807:42)
      at /home/travis/build/hyperledger/composer/node_modules/hfc/node_modules/grpc/src/node/src/client.js:422:14
```

**Comment by davidkel** [comment]
Giving up on this now as it doesn't appear to be stable.

**Pull request closed without merging**